fenix/app/src/main/java/org/mozilla/fenix/compose/ext/Modifier.kt

```diff
@@ -100,3 +115,110 @@ fun Modifier.thenConditional(
     } else {
         this
     }
+
+/**
+ * The Composable this modifier is tied to may appear first and be fully constructed to then be pushed downwards
+ * when other elements appear. This can lead to over-counting impressions with multiple such events
+ * being possible without the user actually having time to see the UI or scrolling to it.
+ */
+private const val MINIMUM_TIME_TO_SETTLE_MS = 1000
+
+/**
+ * Add a callback for when this Composable is "shown" on the screen.
+ * This checks whether the composable has at least [threshold] ratio of it's total area drawn inside
+ * the screen bounds.
+ * Does not account for other Views / Windows covering it.
+ *
+ * @param threshold The ratio of the total area to be within the screen bounds to trigger [onVisible].
+ * @param settleTime The amount of time to wait before calling [onVisible].
+ * @param onVisible Invoked when the UI is visible to the user.
+ */
+fun Modifier.onShown(
+    @FloatRange(from = 0.0, to = 1.0) threshold: Float,
+    settleTime: Int = MINIMUM_TIME_TO_SETTLE_MS,
+    onVisible: () -> Unit,
+): Modifier {
+    val initialTime = System.currentTimeMillis()
+    var lastVisibleCoordinates: LayoutCoordinates? = null
+
+    return composed {
+        if (inComposePreview) {
+            Modifier
+        } else {
+            val context = LocalContext.current
+            var wasEventReported by remember { mutableStateOf(false) }
+
+            val toolbarHeight = context.resources.getDimensionPixelSize(R.dimen.browser_toolbar_height)
```

Comment:
Apologies that I missed this previously, Looking into the tests made me realize about this issue. I'm a bit worried that this modifier code has a lot of dependencies to things it shouldn't directly need. For instance, the common `onShown` modifier shouldn't worry about the rest of the UI hierarchy, in the case of RC bottom sheet, the toolbar placement doesn't even matter since the composable is in a bottom sheet and not a list on home. Wondering if we should use this common logic or think of a different way?

Reply:
I agree that this Modifier has a lot of dependencies and is a good candidate for further refactoring. 

Good catch on the toolbar stuff. If this Modifier is to be reusable, its API will need to be changed to factor something like this in when necessary and not assume it's relevant for 100% of use cases (like review checker). I'll need to think of a good way to accommodate both of these paths, so this will further delay this landing and I may need to hand this off, if it comes to that. I've already spent a good amount of time on this just standing up the requirements of the ticket but also trying to get the UI tests stood up, so I'm hoping this won't take too much longer.